kernel/kernel-api/src/main/java/io/delta/kernel/internal/data/DelegateRow.java

```diff
@@ -0,0 +1,212 @@
+  }
+
+  @Override
+  public int getInt(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      throwIfUnsafeAccess(ordinal, IntegerType.class, "integer");
+      return (int) overrides.get(ordinal);
+    }
+    return row.getInt(ordinal);
+  }
+
+  @Override
+  public long getLong(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      throwIfUnsafeAccess(ordinal, LongType.class, "long");
+      return (long) overrides.get(ordinal);
+    }
+    return row.getLong(ordinal);
+  }
+
+  @Override
+  public float getFloat(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      throwIfUnsafeAccess(ordinal, FloatType.class, "float");
+      return (float) overrides.get(ordinal);
+    }
+    return row.getFloat(ordinal);
+  }
+
+  @Override
+  public double getDouble(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      throwIfUnsafeAccess(ordinal, DoubleType.class, "double");
+      return (double) overrides.get(ordinal);
+    }
+    return row.getDouble(ordinal);
+  }
+
+  @Override
+  public String getString(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      throwIfUnsafeAccess(ordinal, StringType.class, "string");
+      return (String) overrides.get(ordinal);
+    }
+    return row.getString(ordinal);
+  }
+
+  @Override
+  public BigDecimal getDecimal(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      throwIfUnsafeAccess(ordinal, DecimalType.class, "decimal");
+      return (BigDecimal) overrides.get(ordinal);
+    }
+    return row.getDecimal(ordinal);
+  }
+
+  @Override
+  public byte[] getBinary(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      throwIfUnsafeAccess(ordinal, BinaryType.class, "binary");
+      return (byte[]) overrides.get(ordinal);
+    }
+    return row.getBinary(ordinal);
+  }
+
+  @Override
+  public Row getStruct(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      throwIfUnsafeAccess(ordinal, StructType.class, "struct");
+      return (Row) overrides.get(ordinal);
+    }
+    return row.getStruct(ordinal);
+  }
+
+  @Override
+  public ArrayValue getArray(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      // TODO: Not sufficient check, also need to check the element type. This should be revisited
+      //  together with the GenericRow.
+      throwIfUnsafeAccess(ordinal, ArrayType.class, "array");
+      return (ArrayValue) overrides.get(ordinal);
+    }
+    return row.getArray(ordinal);
+  }
+
+  @Override
+  public MapValue getMap(int ordinal) {
+    if (overrides.containsKey(ordinal)) {
+      // TODO: Not sufficient check, also need to check the element type. This should be revisited
+      //  together with the GenericRow.
+      throwIfUnsafeAccess(ordinal, MapType.class, "map");
+      return (MapValue) overrides.get(ordinal);
+    }
+    return row.getMap(ordinal);
+  }
+
+  private void throwIfUnsafeAccess(
+      int ordinal, Class<? extends DataType> expDataType, String accessType) {
+
```

Comment:
nit: you can avoid empty whitespace lines like this at the beginning of functions